src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionPools.cs

```diff
@@ -0,0 +1,346 @@
+                var descriptor = new ServiceDescriptor(serviceName) { HostGroup = hostGroup };
+
+                // call to get service can fail due to this bug - devdiv#288961 or more.
+                // until root cause is fixed, we decide to have retry rather than fail right away
+                for (var i = 0; i < max_retry; i++)
+                {
+                    try
+                    {
+                        // we are wrapping HubClient.RequestServiceAsync since we can't control its internal timeout value ourselves.
+                        // we have bug opened to track the issue.
+                        // https://devdiv.visualstudio.com/DefaultCollection/DevDiv/Editor/_workitems?id=378757&fullScreen=false&_a=edit
+
+                        // retry on cancellation token since HubClient will throw its own cancellation token
+                        // when it couldn't connect to service hub service for some reasons
+                        // (ex, OOP process GC blocked and not responding to request)
+                        return await RetryRemoteCallAsync<OperationCanceledException, Stream>(
+                            () => client.RequestServiceAsync(descriptor, cancellationToken),
+                            timeout,
+                            cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (RemoteInvocationException ex)
+                    {
+                        // save info only if it failed with different issue than before.
+                        if (lastException?.Message != ex.Message)
+                        {
+                            // RequestServiceAsync should never fail unless service itself is actually broken.
+                            // So far, we catched multiple issues from this NFW. so we will keep this NFW.
+                            ex.ReportServiceHubNFW("RequestServiceAsync Failed");
+
+                            lastException = ex;
+                        }
+                    }
+
+                    // wait for retry_delayInMS before next try
+                    await Task.Delay(retry_delayInMS, cancellationToken).ConfigureAwait(false);
+                }
+
+                // crash right away to get better dump. otherwise, we will get dump from async exception
+                // which most likely lost all valuable data
+                FatalError.ReportUnlessCanceled(lastException);
+                GC.KeepAlive(lastException);
+
+                // unreachable
+                throw ExceptionUtilities.Unreachable;
+            }
+
+            #region code related to make diagnosis easier later
+
+            private static readonly TimeSpan s_reportTimeout = TimeSpan.FromMinutes(10);
+            private static bool s_timeoutReported = false;
+
+            private static void ReportTimeout(Stopwatch watch)
+            {
+                // if we tried for 10 min and still couldn't connect. NFW (non fatal watson) some data
```

Comment:
How was 10 minutes picked?

Reply:
just random number picked before we report NFW when connecting to OOP takes long time. this is not actual timeout time.